### PR TITLE
Refactors hydrocarbon combustion + other general atmos code changes

### PR DESCRIPTION
**Pull request by mirrorcult**
basically:

- hydrocarbon combustion is now two separate gas reaction datums, one for tritium combustion and one for plasma combustion. this means that requirement checking is actually sensible and works for both types now
- originally I did this so that we could dynamically populate GLOB.nonreactive_gases rather than hardcoding it, and then I realized that it was kind of impossible to do that considering the conditions (gases that don't react amongst themselves) apply in a lot of different cases, and the hardcoded case of oxygen/nitrogen/CO2 is just for pipelines in their default state. i still added some more gases to the list though. if you can think of a way to dynamically populate the list and keep the current performance buffs introduced, i'd love to hear it.
- also removed two unnecessary max() checks for noblium formation that should've been removed in #37384 
- moves GLOB.nonreactive_gases to gas_types.dm and the gas_reactions singleton to subsystem/air.dm

[why]: theres no reason tritium combustion/plasma combustion should be handled in the same gas reaction datum. it's just snowflakey code and it makes handling the requirements for each a lot more convoluted than they should be. every other reaction type actually makes sense with how it's handled. 

i tested this a little bit and there shouldn't be any difference (except _maybe_ a slight performance boost for react() in some situations?) besides the code being more organized and maintainable

### CODE REVIEW

**Comment by Barhandar** [comment]
TIL plasma and tritium are hydrocarbons.

**Comment by cyclowns** [comment]
they're not TECHNICALLY hydrocarbons but they're "treated as hydrocarbons" in the combustion process

**Comment by ExcessiveUseOfCobblestone** [comment]
who cares it's a fictional gas make it everything at once lol

**Comment by Jalleo** [comment]
I think that the plasma and Tritium is in the same area for performance.

**Comment by cyclowns** [comment]
@Jalleo no, theyre in the same area because super oldcode hydrocarbon combustion used to handle both plasma and "volatile fuel" (which didnt really exist) - it was like this because at one point fire was basically the only reaction, so they needed to be in the same proc. whoever added tritium combustion basically just replaced "volatile fuel" with tritium, reworked the formula a bit and called it a day. tritium/plasma combustion being in the same spot is a remnant of old atmoscode that has no reason to exist in our current gas reaction system.

**Comment by ExcessiveUseOfCobblestone** [comment]
I hope he reports ur post for toxicity lol

**Comment by Jalleo** [comment]
That is a fine enough explanation for me.

File: code/modules/atmospherics/gasmixtures/gas_types.dm [hunk 1]
```diff
@@ -11,11 +12,25 @@ GLOBAL_LIST_INIT(hardcoded_gases, list(/datum/gas/oxygen, /datum/gas/nitrogen, /
...
+		var/datum/gas/gas = g
+		for(var/r in SSair.gas_reactions)
+			if(!gas in nonreactive_gases) //if its been removed already we can move on
+				continue
+			var/datum/gas_reaction/reaction = r
+			if((gas in reaction.min_requirements) || (gas in reaction.max_requirements)) //if the gas is mentioned in the requirements of a reaction
+				nonreactive_gases -= gas
+	. = list()
+	for(var/path in nonreactive_gases)
+		. += new path
```
Comment (AnturK) [inline code/modules/atmospherics/gasmixtures/gas_types.dm]: Why new here ? Not that it matters right now. And no easy way to find out the nonreactive chain is coming to mind, maybe start from the air mix as preset. with that as a base it should be easier to expand.

**Comment by AnturK** [comment]
Looks good to me. Going to throw it for a round or two then merge this.

File: code/modules/atmospherics/gasmixtures/gas_types.dm [hunk 2]
```diff
@@ -11,11 +12,25 @@ GLOBAL_LIST_INIT(hardcoded_gases, list(/datum/gas/oxygen, /datum/gas/nitrogen, /
...
 		gas_info[META_GAS_DANGER] = initial(gas.dangerous)
 		gas_info[META_GAS_ID] = initial(gas.id)
 		.[gas_path] = gas_info
 
+/proc/nonreactive_gas_list() //all the gases not included in _any_ reaction
+	var/list/nonreactive_gases = GLOB.meta_gas_info.Copy()
+	for(var/g in GLOB.meta_gas_info)
+		var/datum/gas/gas = g
+		for(var/r in SSair.gas_reactions)
+			if(!gas in nonreactive_gases) //if its been removed already we can move on
```
Comment (SpaceManiac) [inline code/modules/atmospherics/gasmixtures/gas_types.dm]: Pretty sure the precedence is bad here, coming out to `(!gas) in nonreactive_gases`, but also this proc is never called.

File: code/modules/atmospherics/gasmixtures/reactions.dm [hunk 3]
```diff
@@ -92,16 +92,20 @@
...
+	priority = -1 //fire should ALWAYS be last, but tritium fires happen before plasma fires
+	name = "Tritium Combustion"
+	id = "tritfire"
 
-/datum/gas_reaction/fire/init_reqs()
-	min_requirements = list("TEMP" = FIRE_MINIMUM_TEMPERATURE_TO_EXIST) //doesn't include plasma reqs b/c of other, rarer, burning gases.
+/datum/gas_reaction/tritfire/init_reqs()
+	min_requirements = list(
+		"TEMP" = FIRE_MINIMUM_TEMPERATURE_TO_EXIST,
+		/datum/gas/tritium = MINIMUM_HEAT_CAPACITY,
```
Comment (SpaceManiac) [inline code/modules/atmospherics/gasmixtures/reactions.dm]: Is this the right constant to use here? I thought these were supposed to be mole counts, not heat capacities. Ditto 3 other locations.

**Review by SpaceManiac — CHANGES_REQUESTED** [review]

File: code/modules/atmospherics/gasmixtures/reactions.dm (hunk 3 above)
Comment (AnturK) [reply code/modules/atmospherics/gasmixtures/reactions.dm]: Yeah, even if it's some minimum sensible amount , make a separate define for this please.

File: code/modules/atmospherics/gasmixtures/gas_types.dm (hunk 2 above)
Comment (SpaceManiac) [reply code/modules/atmospherics/gasmixtures/gas_types.dm]: Would prefer to fix or delete it just so if someone does come to use the proc in the future they're not caught off-guard.

Comment (cyclowns) [reply code/modules/atmospherics/gasmixtures/gas_types.dm]: i'll just delete the proc, it is kind of pointless anyway

File: code/modules/atmospherics/gasmixtures/reactions.dm (hunk 3 above)
Comment (cyclowns) [reply code/modules/atmospherics/gasmixtures/reactions.dm]: will do

File: code/controllers/subsystem/air.dm [hunk 4]
```diff
@@ -29,7 +29,8 @@ SUBSYSTEM_DEF(air)
 	var/list/obj/machinery/atmos_machinery = list()
 	var/list/pipe_init_dirs_cache = list()
 
-
+	//atmos singletons
+	var/list/gas_reactions = list()
```
[the remark this answers is not in the document]
Comment (AnturK) [reply code/controllers/subsystem/air.dm]: Can be added in another PR (probably while going over other lists in controllers)

**Comment by duncathan** [comment]
they were kept in the same place in the hopes of an eventual refactor allowing other types of fuel in the burn reaction but this is absolutely fine since that never happened

**Pull request merged**